druntime/src/core/internal/array/capacity.d

```diff
@@ -9,62 +9,150 @@
+ * `_d_arraysetlengthT` and `_d_arraysetlengthiT`, while efficiently handling both
+ * zero-initialized and custom-initialized types.
+ *
+ * ---
+ * ## Example Usage:
+ * ```d
+ * void main()
+ * {
+ *     int[] a = [1, 2];
+ *     a.length = 3; // gets lowered to `_d_arraysetlengthT!(int[])(a, 3)`
+ * }
+ * ```
+ * ---
+ *
+ * - Uses a **templated approach** to minimize `TypeInfo` dependencies.
+ * - Follows `_d_newarrayU` for allocation and initialization.
+ * - Handles memory allocation and resizing safely, ensuring correct initialization.
  */
 
-/// Implementation of `_d_arraysetlengthT` and `_d_arraysetlengthTTrace`
-template _d_arraysetlengthTImpl(Tarr : T[], T)
+/// Complete templated implementation of `_d_arraysetlengthT` and its GC profiling variant `_d_arraysetlengthTTrace`
+
+size_t _d_arraysetlengthT(Tarr : T[], T)(
+    return scope ref Tarr arr,
+    size_t newlength,
+    string file = __FILE__,
+    int line = __LINE__,
+    string func = __FUNCTION__
+) @trusted
 {
-    private enum errorMessage = "Cannot resize arrays if compiling without support for runtime type information!";
+    import core.lifetime : emplace;
+    import core.internal.array.utils : __arrayAlloc;
+    import core.stdc.string : memcpy, memmove;
+    import core.internal.traits : Unqual;
 
-    /**
-     * Resize dynamic array
-     * Params:
-     *  arr = the array that will be resized, taken as a reference
-     *  newlength = new length of array
-     * Returns:
-     *  The new length of the array
-     * Bugs:
-     *   The safety level of this function is faked. It shows itself as `@trusted pure nothrow` to not break existing code.
-     */
-    size_t _d_arraysetlengthT(return scope ref Tarr arr, size_t newlength) @trusted pure nothrow
+    alias U = Unqual!T; // Ensure non-inout type
+
+    if (newlength == 0)
+    {
+        arr = Tarr.init;
+        return 0;
+    }
+
+    static if (is(T == immutable) || is(T == const))
     {
-        version (DigitalMars) pragma(inline, false);
-        version (D_TypeInfo)
+        // Shrink case
+        if (newlength <= arr.length)
         {
-            auto ti = typeid(Tarr);
+            arr = arr[0 .. newlength];
+            return arr.length;
+        }
 
-            static if (__traits(isZeroInit, T))
-                ._d_arraysetlengthT(ti, newlength, cast(void[]*)&arr);
-            else
-                ._d_arraysetlengthiT(ti, newlength, cast(void[]*)&arr);
+        // Allocate new array for immutable/const
+        auto tempArr = new U[newlength];
 
-            return arr.length;
+        // Copy existing elements
+        if (arr.ptr !is null)
+        {
+            memmove(tempArr.ptr, arr.ptr, arr.length * U.sizeof);
+        }
+
+        // Ensure initialization of new elements for immutable types
+        foreach (i; arr.length .. newlength)
+        {
+            emplace(&tempArr[i], U.init);
+        }
+
+        arr = cast(Tarr) tempArr;
+        return arr.length;
+    }
+
```

Comment:
Don't allocate an extra array. The old hook tried to reallocate it using `gc_expandArrayUsed`. Keep the same inner logic.